### PR TITLE
Onload ssr fix

### PR DESCRIPTION
**Pull request by eldar-gamisoniya**

### CODE REVIEW

**Comment by faceyspacey** [comment]
Ah man brother, did you not see my message:

https://github.com/faceyspacey/react-universal-component/issues/30#issuecomment-326123552

We can reduce this to almost a one line change by just not doing caching on the server. 

There's only one place it could be slightly problematic: with double renders such as those Apollo and other solutions use. But that ultimately wouldn't be a problem with `store.replacerReducer` and likely most things you do in `onLoad`. I was fine with just doing that until we heard of any specific problems people find, since that reduces the complexity of implementation by a lot. Also, very few people are even doing the double render thing.

**Comment by eldar-gamisoniya** [comment]
@faceyspacey ah, sorry, missed that one :( . If you currently don't have time to review this i can create new pull request with the temporary fix you mentioned.

But still, this solution solves the other possible problems i mentioned before.
1. cache is global, so if multiple universal components reference the same module, it would be reused from cache both on the server and client, which should improve performance, instead of hurting it on double renders.
2. cache stores modules instead of react components, so if there are dynamic keys it will still work.
3. using context is problematic, as contextTypes currently get only store. While it's possible to solve the problem with onLoad triggering each time on the server, using a library can create a lot of frustration: as the docs say "this option to the universal HOC is only fired the first time the module is received", which is not always true, and if such a situation will be happening only on the server, it will be quite hard to find out the source of the problem. And again, simplifying the onLoad logic create complications for the library users, such as need to make every onLoad function stateful and store some flags in global/module scope (which will happen quite often, for example running module's root saga or creating other side effects should happen only once)

**Comment by faceyspacey** [comment]
let's do it the one line way. also it likely won't be temporary. i have stuff coming out for RFR + Apollo which kills the double render thing. the double render thing is a hack and shouldn't be happening. also the percentage of people doing it is extremely small. I appreciate the time you put in, i wish u saw the note. Let's make the change and move on with our lives.

**Comment by faceyspacey** [comment]
If and when someone has this problem (i.e. 3 weeks from now), they can create a PR for this option:

```js
import universal from 'react-universal-component`

await renderFirstTimeInVirtualDom()
universal.blockOnLoad()

const appString = ReactDOMServer.renderToString(<App />)
```

...at the end of the day adding tons and tons of code to account for an uncommon edge case *that react was never initially intended for and isnt supported by React itself* isn't good for us. I gotta maintain this code. I dont wanna maintain a lot of code for little gain. ...so the above `blockOnLoad` function is different as it's totally separate and also like 1-2 lines of code. It's very *explicit*. It puts control in the user's hands too, which can perhaps smooth over even further edge cases. 

So something like above as an explicit solution directly to the given problem is probably the right way to go. No need to implement it. I don't want to expand the API surface unless we're encountering users with the problem. It may turn out nobody any time soon comes with this problem. For Redux-First Router I promote seamless route-based single-render approach:

https://medium.com/faceyspacey/server-render-like-a-pro-w-redux-first-router-in-10-steps-b27dd93859de

The only reason to need otherwise in my opinion is if you're using Apollo. But I want to bring the best of Apollo to the route level within RFR. I've written about it elsewhere ad nauseum, but i think co-located component-level data fetching is never needed unless ur the size of Facebook and want to make it easy for developers to work independently just at the component level. That's 1 company in the world, and everyone has copied them, while making their lives more complicated. Route level wins 9 out of 10 reasons in my assessment.

File: src/requireUniversalModule.js [hunk 1]
```diff
@@ -51,7 +51,7 @@ export default function requireUniversalModule<Props: Props>(
   const requireSync = (props: Object, context: Object): ?any => {
     let exp = loadFromCache(chunkName, props, modCache)
 
-    if (!exp) {
+    if (!exp || isServer) {
```
Comment (faceyspacey) [inline src/requireUniversalModule.js]: tuck it in `loadFromCache`, as `requireAsync` may need it too.

Also, additional formatting was added to the code that should not have been.

**Comment by eldar-gamisoniya** [comment]
@faceyspacey 
Tried to use onLoad function in test project. The working solution in userland to load the module only once is not that trivial: you can't use context, as in contextTypes there is only store (you can create new reducer, that will handle which modules were loaded, but it won't work by default behaviour, as this state would be rehydrated on the client), you can't create just a global boolean variable in user's code, as it doesn't know about flushing logic (so it won't refresh between requests on the server). So even with this fix, you only able to either have only one universal component per module, or the onLoad function that can be called multiple times (can't run sagas there, for example), or implement custom flushing logic for modules (which seems too hacky for me). 

PS blockOnLoad is not really helps there in case of multiple universal components in module, that will trigger onLoad function multiple times on the client

**Comment by faceyspacey** [comment]
I honestly can't understand. It sounds like ur describing the multi render approach. Does it work for single renders? Why would we need context for single renders?

But yea context currently only has store as it must be specified. Options could be provided to allow for more things.

**Comment by eldar-gamisoniya** [comment]
Use cases:
1. multiple renders of the same universal component on the server: blockOnLoad you mentioned before helps, either way user need to create custom flushing logic for onLoad function.
2. multiple universal components that imports the same module: blockOnLoad won't help as they will be called multiple times during the same render on the server, and will be called more than once (synchronously or asynchronously) on the client as well. Situation with multiple universal components can happen quite often on big projects while leveraging async modules system, as modules in react projects usually meant to store the whole business logic (reducers, sagas, etc) and multiple container components that connected to module's store.

**Comment by faceyspacey** [comment]
I'm having a hard time seeing the second one. When would you want to import the same component/module twice? You only need it once. The only use case I see is if u display it twice in the same page. But that only really happens with things like rows. Maybe I'm missing something.

**Comment by eldar-gamisoniya** [comment]
While splitting your application into per-feature modules, they are not necessarily follow the structure of one component per module, as in the per-page modules. In projects i've been working, modules usually separated by business domains (reducers/sagas), and can have multiple different containers to export (not the rows, but, for example audio player at the top of the page and playlist somewhere else, in the module 'music', with single module reducer and root saga).

**Comment by faceyspacey** [comment]
Got it.

Still code splitting with webpack to begin with us fragile. We can't support every setup. Webpack for example will create chunks for every file in a folder if you are doing the pattern where a folder contains styles, reducer, etc, while using dynamic expressions for your imports.

I want to accommodate as many scenarios as possible, but the scenario you are talking about is very rare for people serious about code splitting. 95% of the time they are able to devote a file to the component or even make a new file that is one line long. That is basically what you have to do if you are doing the above folder organization strategy I just mentioned if you don't want unnecessary chunks created. Or use the context replacement plugin but that's even more to annoying. The recommended strategy is to make an exclusive asyncComponents folder.

People can rearrange their code for code splitting. 

But here's the other part: I will immerse myself in this code in the future. We just something that works now. There's a semi rewrite coming. Other opportunities will open themselves up to us. And I don't have time to introduce anything big right now. In a about a week I'm doing nothing but working on Universal. But for now we need to get out the simple approach. I was gonna do it myself but I wanted to give u a chance to finish it since u put the work in so u would appear as a contributor, etc. I'll be merging this today.

**Pull request merged**